filters/FirstInVoxelFilter.cpp

```diff
@@ -0,0 +1,131 @@
+ * modification, are permitted provided that the following
+ * conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+ *       names of its contributors may be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "FirstInVoxelFilter.hpp"
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{"filters.firstInVoxel",
+                                     "First Entry Voxel Filter", 
+                                     "http://pdal.io/stages/filters.firstInVoxel.html"};
+
+CREATE_STATIC_STAGE(FirstInVoxelFilter, s_info)
+
+std::string FirstInVoxelFilter::getName() const
+{
+    return s_info.name;
+}
+
+void FirstInVoxelFilter::addArgs(ProgramArgs& args)
+{
+    args.add("cell", "Cell size", m_cell, 0.001);
+}
+
+PointViewSet FirstInVoxelFilter::run(PointViewPtr view)
+{
+    PointViewPtr output = view->makeNew();
+    for (PointId id = 0; id < view->size(); ++id)
+    {
+        if (voxelize(view->point(id)))
+        {
+            output->appendPoint(*view, id);
+        }
+    }
+
+    PointViewSet viewSet;
+    viewSet.insert(output);
+    return viewSet;
+}
+
+bool FirstInVoxelFilter::voxelize(const PointRef point)
+{
+    /*
+     * Calculate the voxel coordinates for the incoming point.
+     * gx, gy, gz will be the global coordinates from (0, 0, 0).
+     */
+    int gx = point.getFieldAs<double>(Dimension::Id::X) / m_cell;
+    int gy = point.getFieldAs<double>(Dimension::Id::Y) / m_cell;
+    int gz = point.getFieldAs<double>(Dimension::Id::Z) / m_cell;
+
+    static bool initialized = false;
```

Comment:
This needs to be moved to be a member variable or multiple instances of this filter will be confused.

Reply:
Yeah, make sense.
Changed to member variable